src/tools/crossgen2/Common/TypeSystem/Common/ExplicitLayoutValidator.cs

```diff
@@ -0,0 +1,146 @@
+        private enum FieldLayoutTag : byte
+        {
+            Empty,
+            NonORef,
+            ORef,
+        }
+
+        private readonly int _pointerSize;
+
+        private readonly FieldLayoutTag[] _fieldLayout;
+
+        private readonly MetadataType _typeBeingValidated;
+
+        private ExplicitLayoutValidator(MetadataType type, int typeSizeInBytes)
+        {
+            _typeBeingValidated = type;
+            _pointerSize = type.Context.Target.PointerSize;
+            _fieldLayout = new FieldLayoutTag[typeSizeInBytes];
+        }
+
+        public static void Validate(MetadataType type, ComputedInstanceFieldLayout layout)
+        {
+            ExplicitLayoutValidator validator = new ExplicitLayoutValidator(type, layout.ByteCountUnaligned.AsInt);
+            foreach (FieldAndOffset fieldAndOffset in layout.Offsets)
+            {
+                validator.AddToFieldLayout(fieldAndOffset.Offset.AsInt, fieldAndOffset.Field.FieldType);
+            }
+        }
+
+        private void AddToFieldLayout(int offset, TypeDesc fieldType)
+        {
+            if (fieldType.IsGCPointer)
+            {
+                if (offset % _pointerSize != 0)
+                {
+                    // Misaligned ORef
+                    ThrowFieldLayoutError(offset);
+                }
+                SetFieldLayout(offset, _pointerSize, FieldLayoutTag.ORef);
+            }
+            else if (fieldType.IsValueType)
+            {
+                MetadataType mdType = (MetadataType)fieldType;
+                int fieldSize = mdType.InstanceByteCountUnaligned.AsInt;
+                if (!mdType.ContainsGCPointers)
+                {
+                    // Plain value type, mark the entire range as NonORef
+                    SetFieldLayout(offset, fieldSize, FieldLayoutTag.NonORef);
+                }
+                else
+                {
+                    if (offset % _pointerSize != 0)
+                    {
+                        // Misaligned struct with GC pointers
+                        ThrowFieldLayoutError(offset);
+                    }
+
+                    bool[] fieldORefMap = new bool[fieldSize];
+                    MarkORefLocations(mdType, fieldORefMap, offset: 0);
+                    for (int index = 0; index < fieldSize; index++)
+                    {
+                        SetFieldLayout(offset + index, fieldORefMap[index] ? FieldLayoutTag.ORef : FieldLayoutTag.NonORef);
+                    }
+                }
+            }
+            else if (fieldType.IsPointer || fieldType.IsByRef || fieldType.IsByRefLike)
+            {
+                if (offset % _pointerSize != 0)
+                {
+                    // Misaligned pointer field
+                    ThrowFieldLayoutError(offset);
+                }
+                SetFieldLayout(offset, _pointerSize, FieldLayoutTag.NonORef);
+            }
+            else
+            {
+                throw new NotImplementedException();
```

Comment:
Nit: Instead of an else branch that throws NotImplemented, I tend to use asserts. It doesn't make sense to ship the throw to customers and if we ever add new type system entities, the assert will hit.

Reply:
Changed NotImplementedException throw to assertion failure.